services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt

```diff
@@ -32,6 +58,264 @@ class HakemusService(
+        logger.info("Updating application id=$applicationId")
+
+        val applicationEntity =
+            applicationRepository.findOneById(applicationId)
+                ?: throw ApplicationNotFoundException(applicationId)
+        val application = applicationEntity.toApplication() // the original state for audit logging
+
+        assertNotSent(applicationEntity)
+        assertCompatibility(applicationEntity, request)
+
+        if (!request.hasChanges(applicationEntity)) {
+            logger.info(
+                "Not updating unchanged application data. id=$applicationId, identifier=${applicationEntity.applicationIdentifier}"
+            )
+            return hakemusResponseWithYhteystiedot(applicationEntity)
+        }
+
+        assertGeometryValidity(request.areas) { validationError ->
+            "Invalid geometry received when updating application id=${applicationEntity.id}, reason = ${validationError.reason}, location = ${validationError.location}"
+        }
+
+        assertYhteystiedotValidity(applicationEntity, request)
+
+        val hankeEntity = applicationEntity.hanke
+        if (!hankeEntity.generated) {
+            request.areas?.let { areas ->
+                assertGeometryCompatibility(hankeEntity.id, areas) { area ->
+                    "Application geometry doesn't match any hankealue when updating application id=${applicationEntity.id}, " +
+                        "hankeId = ${hankeEntity.id}, application geometry = ${area.geometry.toJsonString()}"
+                }
+            }
+        } else {
+            updateHankealueet(hankeEntity, request)
+        }
+
+        val updatedApplicationEntity = saveWithUpdate(applicationEntity, request)
+
+        logger.info("Updated application id=${applicationId}")
+        applicationLoggingService.logUpdate(
+            application,
+            updatedApplicationEntity.toApplication(),
+            userId
+        )
+
+        return hakemusResponseWithYhteystiedot(updatedApplicationEntity)
+    }
+
+    /** Assert that the application has not been sent to Allu. */
+    private fun assertNotSent(applicationEntity: ApplicationEntity) {
+        if (applicationEntity.alluid != null) {
+            throw ApplicationAlreadySentException(applicationEntity.id, applicationEntity.alluid)
+        }
+    }
+
+    /** Assert that the update request is compatible with the application data. */
+    private fun assertCompatibility(
+        applicationEntity: ApplicationEntity,
+        request: HakemusUpdateRequest
+    ) {
+        when (applicationEntity.applicationData) {
+            is CableReportApplicationData ->
+                if (request !is JohtoselvityshakemusUpdateRequest) {
+                    throw IncompatibleHakemusUpdateRequestException(
+                        applicationEntity.id!!,
+                        applicationEntity.applicationData::class,
+                        request::class
+                    )
+                }
+        }
```

Comment:
This will be easier to extend if checking the condition and throwing the exception are separated. There's also less nesting this way, which is better for our brains.
```suggestion
        val expected =
            when (applicationEntity.applicationData) {
                is CableReportApplicationData -> request is JohtoselvityshakemusUpdateRequest
            }
        if (!expected) {
            throw IncompatibleHakemusUpdateRequestException(
                applicationEntity.id!!,
                applicationEntity.applicationData::class,
                request::class
            )
        }
```